torchtune/modules/transformer.py

```diff
@@ -150,6 +157,9 @@ def __init__(
         self.head_dim = head_dim
         self.causal_mask = None
 
+        self.layer_dropouts = create_layer_dropout_modules(num_layers, layer_dropout_prob, layer_dropout_prob_layer_scale, layer_dropout_str)
+        self.output_hidden_states = OrderedDict() # TODO: use tensordict?
```

Comment:
Just curious: why do we need a dict at all? And separately, why is this defined as an instance variable instead of something local to forward? Is it because you don't want to change the return signature? (If so I think we should just change it since returning hidden states is a fairly common case anyways)

Reply:
I made hidden_states local to forward() and returned in this commit:
ccb4a5006ece5436505d8f7b1764734aa347d422

One disadvantage with returning hidden states is that it requires adding `if..else` conditions in the return statement and in the training script. If you prefer that, should I create a dedicated training script instead? I was hoping though that early exit and layer dropout become orthogonal to all of the existing training scripts: single node, multi node, lora, etc.

Reply:
> why do we need a dict at all?

It is for a couple of reasons:
- we want different layers to have different loss scales (I implemented that in a separate function)
- I want to give the user flexibility which layers to return hidden state (in an upcoming commit or PR I might even have different iterations have different layers outputting hidden states)

Reply:
Resolving as `output_hidden_states` got supported in another PR.